File: l33tcode/plus-one-linked-list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def plusOne(self, head: ListNode) -> ListNode:
        if not head:
            return ListNode(1)

        def reverse_list(head: ListNode) -> ListNode:
            prev, cur = head, head.next
            prev.next = None

            new_head = prev

            while cur:
                new_head = cur
                cur_next = cur.next
                cur.next = prev
                prev, cur = cur, cur_next

            return new_head

        new_head = reverse_list(head)

        node = new_head

        carry = 1
        while node and carry > 0:
            node.val, carry = (node.val + carry) % 10, (node.val + carry) // 10
            node = node.next

        new_head = reverse_list(new_head)

        if carry > 0:
            carry_head = ListNode(1)
            carry_head.next = new_head
            new_head = carry_head

        return new_head

File: l33tcode/plus-one-linked-list.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
private:
  ListNode *reverseList(ListNode *head) {
    auto node = head;
    ListNode *prevNode = nullptr;

    while (node) {
      auto curNode = node;
      auto nextNode = node->next;

      node = nextNode;
      curNode->next = prevNode;
      prevNode = curNode;
    }

    return prevNode;
  }

public:
  ListNode *plusOne(ListNode *head) {
    auto newHead = new ListNode(0, head);
    auto reversedListHead = reverseList(newHead);

    int carry = 1;

    auto node = reversedListHead;

    while (node && carry) {
      int val = node->val;
      node->val = (val + carry) % 10;
      carry = (val + carry) / 10;
      node = node->next;
    }

    reverseList(reversedListHead);

    return newHead->val > 0 ? newHead : newHead->next;
  }
};

int main() { return 0; };
